RecoTauTag/RecoTau/interface/PFRecoTauClusterVariables.h

```diff
@@ -24,39 +60,82 @@ namespace reco { namespace tau {
   float eratio(const pat::Tau& tau);
   /// return sum of pt weighted values of distance to tau candidate for all pf photon candidates, 
   /// which are associated to signal; depending on var the distance is in 0=:dr, 1=:deta, 2=:dphi 
-  float pt_weighted_dx(const reco::PFTau& tau, int mode = 0, int var = 0, int decaymode = -1);
-  float pt_weighted_dx(const pat::Tau& tau, int mode = 0, int var = 0, int decaymode = -1);
+  template<class TauType, class PhotonVectorType>
+  float pt_weighted_dx(const TauType& tau, int mode = 0, int var = 0, int decaymode = -1) {
+    float sum_pt = 0.;
+    float sum_dx_pt = 0.;
+    float signalrad = std::max(0.05, std::min(0.1, 3./std::max(1., tau.pt())));
+    int is3prong = (decaymode==10);
+    const auto& cands = getGammas<TauType, PhotonVectorType>(tau, mode < 2);
+    for (const auto& cand : cands) {
+      // only look at electrons/photons with pT > 0.5
+      if (cand->pt() < 0.5){
+        continue;
+      }
+      float dr = reco::deltaR(cand->eta(), cand->phi(), tau.eta(), tau.phi());
+      float deta = std::abs(cand->eta() - tau.eta());
+      float dphi = std::abs(reco::deltaPhi(cand->phi(), tau.phi()));
+      float pt = cand->pt();
+      bool flag = isInside(pt, deta, dphi);
+      if(is3prong==0){
+        if (mode == 2 || (mode == 0 && dr < signalrad) || (mode == 1 && dr > signalrad)) {
+          sum_pt += pt;
+          if (var == 0)
+            sum_dx_pt += pt * dr;
+          else if (var == 1)
+            sum_dx_pt += pt * deta;
+          else if (var == 2)
+            sum_dx_pt += pt * dphi;
+        }
+      }
+      else if(is3prong==1){
+        if( (mode==2 && flag==false) || (mode==1 && flag==true) || mode==0){
+          sum_pt += pt;
+          if (var == 0)
+            sum_dx_pt += pt * dr;
+          else if (var == 1)
+            sum_dx_pt += pt * deta;
+          else if (var == 2)
+            sum_dx_pt += pt * dphi;
+        }
+      }
+    }
+    if (sum_pt > 0.){
+      return sum_dx_pt/sum_pt;  
+    }
+    return 0.;
+  }
   /// return sum of pt weighted values of dr relative to tau candidate for all pf photon candidates,
   /// which are associated to signal
   inline float pt_weighted_dr_signal(const reco::PFTau& tau, int dm) {
-    return pt_weighted_dx(tau, 0, 0, dm);
+    return pt_weighted_dx<reco::PFTau, const std::vector<reco::PFCandidatePtr>&>(tau, 0, 0, dm);
```

Comment:
I see now how this can be semantically more complex.
My version did not require the rather lengthy template parameter specifications in every call, which seems rather tedious/verbose.
It makes more sense to not expose the templated methods to the users in the interface, which can only really work for two specific pairs of types.

Reply:
I agree that your proposal is nice as leads to more compact code and as it builds correct pairs of types. I understand that you request to further modify the code => it will be done